# Problem 38 - Pandigital multiples

Find the largest 9-digit pandigital concatenated product of an integer and a sequence 1, 2, ..., n where n > 1

def: a pandigital number is a number that includes all the digits 1-9

def: a concatenated product of an integer x and a tuple (a, b, c) is the concatenation x * a + x * b + x * c in string form

# Tinkering

In [4]:
def is_pandigital(n):
    digits = ['1', '2', '3', '4', '5', '6', '7', '8', '9']
    
    n  = str(n)
    
    is_pd = True
    
    for d in digits:
        if d not in n:
            is_pd = False
            break
            
    return is_pd

In [5]:
# testing
is_pandigital(123456789)

True

In [6]:
is_pandigital(1)

False

In [7]:
def concat_product(x, tup):
    """
    returns the concatenated product of int x and tuple tup
    """
    
    prod_str = ''
    
    for tup_elem in tup:
        part_str = str(x * tup_elem)
        prod_str += part_str
        
    return prod_str

In [9]:
# testing
concat_product(9, (1,2,3, 4, 5))

'918273645'

In [10]:
## random guessing
concat_product(99, (1,2,3))

'99198297'

In [11]:
concat_product(90, (1,2,3,4))

'90180270360'

In [12]:
concat_product(999, (1, 2))

'9991998'

In [13]:
concat_product(93, (1,2,3,4))

'93186279372'

In [14]:
concat_product(9396, (1,2))

'939618792'

In [15]:
for x in range(9000, 10000):
    if is_pandigital(concat_product(x, (1,2))):
        print(x, concat_product(x, (1,2)))

9267 926718534
9273 927318546
9327 932718654


## Reasoning for solution

In the example given, 918273645 is obtained. The highest pandigital concat sum should be larger.

This can only be true if the first digit of the integer in the concatenated sum is 9.

There is no tuple that gives a 9-digit concatenated sum for a two digit or three digit integer starting with 9

Thus the integer must have four digits.

This narrows down the number of integers that need to be tested to only a thousand. 

This is fairly quick.

# Solution:
The concatenated sum of 9327 and (1,2) is 932718654